centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+    
+    - `obj->Mode : int`: A bitmask of the O_* flags, which include: `O_RDONLY` (read only), `O_WRONLY` (write only), `O_RDWR` (read/write), `O_CREAT` (create), `O_TRUNC` (truncate), and `O_EXCL` (exclusive, see above).
+    
+    - `obj->Pathname : pPathname`: A pointer to a Pathname struct (defined in `include/obj.h`) which contains the complete parsed pathname for the object.  This provides a buffer for the pathname as well as an array of pointers to the pathname's components.  The function `obj_internal_PathPart()` can be used to obtain at will any component or series of components of the pathname.
 
-Finally, if the os driver specified a capability of OBJDRV_C_TRANS, it must pay attention to the current state of the end-user's trans- action.  If a new object is being created, an object is being deleted, or other modifications/additions are being performed, and if the OXT layer indicates a transaction is in process, the driver must either complete the current transaction and then complete the current call, or else add the current delete/create/modify call to the transaction tree (in which case the tree item is preallocated; all the driver needs to do is fill it in).  The transaction layer will be discussed in depth later in this document.
+    - `obj->Pathname->OpenCtl : pStruct[]`: Parameters for the open() operation, as defined by the driver author. These are specified in the path in a similar way to URLs (`example.qy?param1=value&param2=other_value`).  Drivers typically only use `obj->Pathname->OpenCtl[obj->SubPtr]` (see SubPtr below) to retrieve their own parameters, ignoring parameters passed to other drivers in the path.
 
-As a part of the Open process, the OS driver will normally allocate an internal structure to represent the current open object, and will return that structure as a void* data type in the return value.  This pointer will be then passed to each of the other driver entry point functions, with the exception of QueryFetch, QueryDelete, and Query- Close, which will be discussed later.
+    - `obj->SubPtr : short`: The number of components in the path that are a part of the path to the root node object, including the `.` for the top level directory.  For example, in the above path of `/data/file.csv`, the path would be internally represented as `./ data/ file.csv`, so SubPtr is 3.
 
-The Open() routine is called with five parameters:
+    - `obj->SubCnt : short`: _The driver should set this value_ to show the number of components it controls.  This includes the root node object, so `SubCnt` will always be at least 1.  For example, when opening `/data/file.csv/rows/1`, the CSV driver will read the `SubPtr` of 3 (see above), representing `./ data/ file.csv`. It will then set a `SubCnt` of 3, representing that it will control `file.csv /rows /1`.  (The driver only sets `SubCnt`, `SubPtr` is provided.)
```

Comment:
It'll be good to explicitly state that the node object is included in the counts of both SubPtr and SubCnt.

Reply:
Section updated.